### ISSUE TITLE
Incorporating in fast permutation tests

### ISSUE THREAD

**Issue by mortonjt** [scipy/scipy#4399]
Recently, I've been developing fast permutation tests that can make use of numpy/scipy matrix operations.

So far, I have developed 3 permutation tests: a mean test, a t-test and a pearson correlation test.

To get an idea for the timings, I've benchmarked the mean test to a naive implementation on a 60x60 table and  with 1000 permutations. Below are the timings

Naive time [s]: 1.91008838018
Numpy time [s]: 0.0174480279287

If you guys are down, I can start committing these algorithms into scipy. In the near future, I plan on implementing ANOVA tests such as an F-test and a Krustal-Wallis test.

**Comment by rgommers** [comment]
@mortonjt I don't quite understand what the difference is between the tests with the same names that are already in ``scipy.stats``. Can you elaborate? And do you have them up on Github somewhere already?

**Comment by mortonjt** [comment]
I really appreciate the response.

I've add the code a simple test case to the following code
https://github.com/mortonjt/scipy/blob/master/scipy/stats/stats.py
https://github.com/mortonjt/scipy/blob/master/scipy/stats/tests/test_stats.py

The basic idea is it'll calculate a t-test using permutations rather than assuming normality.
In addition, it can calculate multiple t-tests at once for multiple vectors of the same length given a categorical vector. By making use of numpy matrix multiplication, this is much much faster than looping through and calculating t-statistics for each individual vector.

If the math isn't quite making sense, I'll be happy to add some latex markup in the documentation.

```
def ttest_perm(a, b, axis=0, permutations = 1000):
    """
    Calculates the Welch's T-test for the means of TWO INDEPENDENT samples of scores
    using permutation methods
    This test is an equivalent to scipy.stats.ttest_ind, except it doesn't require the
    normality assumption since it uses a permutation test.
    Parameters
    ----------
    a : array_like
        The array must be 2 dimensional
    b: array_like
        This array must be 1 dimensional and have the same size as
        the length of the specified axis
    axis : int, optional
        Axis can equal None (ravel array first), or an integer (the axis
        over which to operate on a and b).
    permutations: int
        Number of permutations used to calculate p-value
    Returns
    -------
    t : float or array
        The calculated t-statistic.
    prob : float or array
        The two-tailed p-value.
    """
    assert len(a.shape) == 2
    r, c = a.shape
    mat = np.matrix(a)
    if axis == 0: # Test with respect to rows
        assert len(b) == c
    else:
        assert len(b) == r
        mat = mat.transpose()

    perms = _init_categorical_perms(b, permutations=1000)
    perms = np.matrix(perms)
    num_cats = 2
    _, c = perms.shape
    permutations = (c - num_cats) / num_cats
    
    ## Perform matrix multiplication on data matrix
    ## and calculate sums and squared sums
    _sums  = mat * perms
    _sums2 = np.multiply(mat, mat) * perms

    ## Calculate means and sample variances
    tot =  perms.sum(axis = 0)
    _avgs  = _sums / tot
    _avgs2 = _sums2 / tot
    _vars  = _avgs2 - np.multiply(_avgs, _avgs)
    _samp_vars =  np.multiply(tot, _vars) / (tot-1)
    
    ## Calculate the t statistic
    idx = np.arange(0, (permutations+1)*num_cats, num_cats, dtype=np.int32)
    denom  = np.sqrt(_samp_vars[:, idx+1] / tot[:,idx+1]  + _samp_vars[:, idx] / tot[:,idx])
    t_stat = np.divide(_avgs[:, idx] - _avgs[:, idx+1], denom)
    ## Calculate the p-values
    cmps =  abs(t_stat[:,1:]) >= abs(t_stat[:,0])
    pvalues = (cmps.sum(axis = 1) + 1.)/(permutations + 1.)
        
    return map(np.array, map(np.ravel, [t_stat[:, 0],pvalues]))
```

**Comment by rgommers** [comment]
@josef-pkt @jseabold you guys have a better overview of the t-test jungle and whether this fits in ``scipy.stats``. What do you think?

**Comment by rgommers** [comment]
If we want to add all the proposed tests, we'll also need to think of a sane API, to prevent the many ``ttest_xxx`` functions situation for the other ones.

**Comment by rgommers** [comment]
@mortonjt for a 60x60 matrix the calculation time isn't too bad, but how does it scale with rows/cols?

**Comment by rgommers** [comment]
You should avoid using ``np.matrix``, using normal arrays (and matrix multiplication with ``np.dot``) is very much preferable.

**Comment by mortonjt** [comment]
It actually scales pretty well.  

Here are some timings that I have conducted on a 12 core Xeon processor (using multithreaded OpenBLAS)

Here are some of the timings scaling the number of permutations.  N = rows, M = columns

| N      | M        | Permutations  | ttest_ind | numpy
|--------|----------|--------------------|-------------|---------|
| 500   | 13.15  | 300                 | 13.15      | 0.18
| 500   | 26.42  | 600                 | 26.42      | 0.36
| 500   | 39.21  | 900                 | 39.21      | 0.52

And here are some timings scaling the rows (N).  

 N      |  M       |  Permutations | Numpy  
--------|----------|--------------------|------------|
 1000 | 1000   | 5000               |  26.902  
 2000 | 1000   | 5000               |  53.137
 3000 | 1000   | 5000               |  79.091

**Comment by dengemann** [comment]
This looks really interesting.

> If we want to add all the proposed tests, we'll also need to think of a sane API, to prevent the many ttest_xxx functions situation for the other ones.

couldn't one add a paremter to switch between classical and permutation variants instead of inflating the API?

**Comment by josef-pkt** [comment]
Alexandre Gramford has written permutation t-test for mne
https://github.com/mne-tools/mne-python/blob/master/mne/stats/permutations.py
that I was targeting for inclusion in statsmodels

Since this is a new area for scipy.stats, it requires some upfront design decisions, especially given the deprecation policy of scipy.  I don't have a strong opinion yet.

- permutation and bootstrap test will be necessary for python either in scipy.stats or statsmodels or both
- efficient permutation needs to be build into the specific functions to make it efficient, e.g. save on unnecessary repeated calculations. This should be considerably faster than permutation or bootstrap "from the outside". I have an example script for another test somewhere in a scipy issue where we had this discussion before. (see motivation of this issue)
- I think the API could avoid inflating the number of functions by adding a `use_permutation` (boolean) and `permutation_kwds` to the current functions for hypothesis tests. But then it would need to offer the other options also. (Actual implementation could also be in a separate private function.)
- Sticking to the current pattern of returning (statistic, pvalue) limits the possibility of providing additional returns which might reuse existing bootstrap or permutation results

for statsmodels

- this will be added -- eventually
- I haven't worked on statsmodels.stats in a while except for hypothesis tests related to estimation models and for maintenance. Unless there is a PR for statsmodels, I would expect this to get added in two versions or so.
- The advantage of statsmodels is that we are not restricted to simple functions with simple returns. Design and refactoring are, however, still unfinished.
- I don't mind code duplication between statsmodels and scipy.stats in some areas. scipy.stats with the basic functions and statsmodels with the "verbose" functions and with extra frills. (I was thinking of copying and duplicating some statistical tests from scipy.stats once we have decided on a "verbose" return/results class for basic hypothesis tests.) 
- statsmodels has joblib as optional dependency which makes it easier to offer a parallel option, same as Alexandre's version.
- extending permutation and bootstrap beyond the simple cases requires supporting infrastructure (for example iterators as for cross-validation in scikit-learn)


In general, t-test is pretty robust to different distributions, some of the other hypothesis test would benefit more from a permutation version. The same pattern should work for all the two or k-sample tests.

To avoid a vaporware issue: Eventually having it in statsmodels doesn't prevent from adding it to scipy.stats. (But I would also be happy to review a PR for statsmodels. :)

**Comment by josef-pkt** [comment]
correction: The permutation t-test in the above mne link is for one-sample or paired sample. I thought Alexandre and I also discussed two sample tests but maybe I don't remember correctly.

**Comment by josef-pkt** [comment]
as an example for the reuse of the permutation or bootstrap: For the model based t-test I have been using confidence interval options in statsmodels.
Many fields try to get researchers away from an exclusive emphasis on p-values towards reporting effect sizes and confidence intervals.

**Comment by mortonjt** [comment]
Thanks for the insight @josef-pkt !

Its pretty funny that someone came up with this idea independently of me.  The code I developed is a two-sample t-test, but can be multithreaded through the multithreaded OpenBLAS library.  

I'd be down to submit a PR for statsmodel.  You do raise a good point about calculating confidence intervals.

@dengemann - I don't think adding a permutation bool parameter to ttest_ind should be a problem.

**Comment by dengemann** [comment]
Just one note. An implementation that's close to the numpy API with support for arrays inputs for the independent variable would be very nice. This is exactly what we need for analysis of neuroscience data where you have thousands of simultaneous observations that are temporally / spatially correlated. Maybe statsmodels could then use that API internally. 

FYI @josef-pkt the permutation clustering test  that we use in [MNE-Python](https://github.com/mne-tools/mne-python) is a different beast, as I see it after reading @mortonjt examples. It only uses the t-test, a standard f-test (or a repeated measures ANOVA (which we still haven't ported to statsmodels ...)) as arbitrary score that's used to bind observations together and for weighting the. It actually does 3 things in 1 step, computing a non-parametric test statistic, controlling the family-wise error rate, and mitigating the multiple comparisons problem (by lumping together adjacent observations via clustering).
The actual test statistic is not a t-test, nor an F-test, etc. but typically the cluster sizr or the sum of all scores in the cluster. The distribution is then computed from recomputing the clustering and the test statistic after permuting the condition labels. A spatial/temporal/spectral connectivity prior is used in addition to allow clustering of adjacent sensors / dipoles in the brain model for adjacent time points and frequency bands. See http://www.ncbi.nlm.nih.gov/pubmed/17517438.

**Comment by josef-pkt** [comment]
@dengemann   Can you be a bit more specific about what would be a function (in scipy.stats style) that would be useful for you and mne?

@mortonjt  specifically about parallelization: this needs some performance benchmarks to see what actually works. I'm skeptical about low level Blas/Lapack parallelization unless they are big matrix/array/linalg calculations. I don't have clear numbers, but someone reported that turning off multithreaded Blas/Lapack made the code actually faster. Additionally, on Windows (what I'm using) the calculations need to take at least 5 to 10 seconds to compensate for the "forking" cost.

Some more about possible features and requirements:

Some users or use cases only need the basic permutation version with fixed number of bootstrap replication as replacement for the usual p-values. Any additional complications or complexities are useless overhead.

The opposite case is when the user wants "reliable" bootstrap results and possibly results additionally to the p-values, and might not not know for sure in advance what he or she wants. (I was thinking more about bootstrap but most of it will also apply to permutation.)

example: adaptive number of bootstrap/permutation samples. If the p-value is around 0.7, then 0.1 up or down doesn't make a difference. If the p-value is 0.045, then the user might want to get less noisy p-values. If the relevant p-value threshold for the hypothesis test is 0.001, then we would need many replications to get good precision in that neighborhood. solution: automatic adaptive stopping rules or give the user the option to continue the previous bootstrap/permutation runs.

example: which results
If the user knows in advance which results he or she wants, e.g. percentile p-value plus corresponding confidence intervals for given confidence level, then we can calculate this and throw away the individual permuation/bootstrap results.
The other case is when users decide what they would like to look at after seeing the first results. In this case the best would be to calculate several statistics from the bootstrap/permutation samples, store them and then calculate additional results on demand.

For simple cases like t-test the full version might be overkill in many cases, for more difficult cases like bootstrap in regression based tests, I can think of many options that I would be interested in as user and that require more flexibility than in the simple case.
The simple case can be implemented in a function in the style of scipy.stats, the second version would need to, optionally, hold on to the current state of the bootstrap/permutations and would be better implemented in a class in the style of statsmodels.

**Comment by dengemann** [comment]
>  Can you be a bit more specific about what would be a function (in scipy.stats style) that would be useful for you and mne?
@josef-pkt sorry, yes. Very simple, all testst in scipy.stats that take arrays of the shape (n_samples, n_features) and returns an array of shape (n_features), are super neat. We typically have 10-300 samples, depending on whether you analyze single subejct or group data but easily end up with one million observations features. Second we try to avoid dependencies. Therefore haveing a test in scipy without fancy API (returning summary objects, etc) is very useful.

**Comment by mortonjt** [comment]
@dengemann that is exactly the kind of data that I this permutation test was originally intended for.  We have matrices with 1000s of samples and 1000s of features.  However, if we wanted to conform to the existing ttest_ind design, a groupby operation would need to be performed to split the samples into two separate categories.

@josef-pkt , I've actually seen the vice versa.  When I enabled multithreaded OpenBLAS on Macbook, I saw a speed up ~50x.  I'll conduct a more rigorous benchmark in a bit.

**Comment by mortonjt** [comment]
@josef-pkt, I really like these ideas.  However, I'm a little worried about blowing up the complexity of the t-test within scipy.stats.  Would you be down to continue this thread over in statsmodels once this modified t-test has been merged?

**Comment by josef-pkt** [comment]
> Would you be down to continue this thread over in statsmodels once this modified t-test has been merged?

Yes,

Just another thought (that I just remembered).  t-test is relatively robust to deviation from normality if the deviation is not "huge". 
However it's sensitive to outliers, heavy tails and very skewed distributions. I don't know how permutation p-values behave in those cases, but one solution for a robust t-test is to use trimming and some of those tests would benefit from a permutation or bootstrap version, IIRC.

another case: testing equality of variances (one, 2 or k-sample) is **very** sensitive and non-robust to deviations from normality. 

and one more: in statsmodels I added an alternative (2-sided, larger, smaller) option to the t- and similar tests. Your permutation version (in the PR) uses the two-sided alternative (if I read/skimmed it correctly). While we can recover the one sided p-values in the standard case using the t-distribution, the one-sided p-value in the permutation test might be rather different from the two-sided if the distribution is skewed.  (I never tried this out.)

(again the two extremes: coding for the most common, simple use case or going for the whole enchilada :)

**Comment by sturlamolden** [comment]
Technically the t-test does not require the data to be normally distributed. It requires the sample mean to be normally distributed. This is a common misunderstanding. I reckon at least 2 of 3 scientists who regularly do statistics do not understand this or what it means. The common notion that we should only the t-test on normally distributed data is just BS. Because of the central limit theorem we know that the sample mean is asymptotically normally distributed. So with a certain value of n, typically 25 or so, the assumptions of the t-test can (almost) always be considered valid. The only exception would be data with some strange discrete distribution.

**Comment by sturlamolden** [comment]
Since I am a Bayesian at heart I would also like to add that a "two-sided" test against a sharp null-hypotehesis is dubious business. We know the null hypothesis to be false *a priori*, it's just a matter of collecting enough data. Two populations will in practive never have exactly the same expectancy value. So why test for something we already know to be false?

**Comment by sturlamolden** [comment]
As for BLAS: On OSX 10.9 and 10.10, Accelerate Framework is usually faster than OpenBLAS and MKL, at least in my own tests. In earlier OSX versions Accelerate Framework is considerably slower. So depending on OSX version you can expect a speedup or slowdown from going to OpenBLAS.

**Comment by mortonjt** [comment]
All good points.

But what if you don't have enough data?  The data that I typically deal with in extremely sparse and skewed, so t-tests really don't work very well.

Glad you did those benchmarks.  Good thing that you can compile the Accelerate Framework/MKL with numpy :)

**Comment by sturlamolden** [comment]
My main issue is with the p-value and testing og sharp null hypotheses, so I do not consider permutation tests to answer a question I am interested in. If I were to use resampling methods I would use the bootstrap (probably BCa) and focus on the effect size. For the same reason I would compute confidence intervals instead of testing hypotheses if I were to use classical parametric statistics. 

I would like to say that the distribution of the data is rarely unknown. It can almost always be worked out from the stochastic processes by which the data is generated and sampled. Scientists who *think they need to use non-parametric statistics* should take it as an indication that they *ought to be consulting a professional statistican*. It is almost always due to lack of knowledge about mathematical statistics and stochastic processes.

In Bayesian statistics we also have a tool called "Markov Chain Monte Carlo",  which allows us to work with even the most untractable distributions. It is computer-intensive methods like the permutation test and the bootstrap in classical statistics, but we consider the data to be fixed and the parameter to be random. So instead of resampling the data we simulate the random parameter.

**Comment by sturlamolden** [comment]
**Failure to reject a null-hypotesis only tells me I did not collect enough data, because I already know a sharp null-hypotesis cannot be true.** So the interesting information as far as I am concerned is the effect-size. How big is it, and is it **big enough to be relevant for whatever I am studying?** For example the Donnan potential for a living cell is about -2 mV. And if I study the membrane potential of neurons there will be signals in the range of -100 to +200 mV, and so the contribution from the Donnan potential is not very relevant. I can collect data and prove the Donnan potential in neurons is *very significant*, but it does not make it very *important*. The word *significance* conveys some sort of information that we should consider it *important*, but that is not a valid interpretation. Particularly in fields like social medicine, epidemiology, human geography, sociology you will see statistics abused this way. "Some factor X gives a significant risk for a certain disease," and media loves to write about this kind of research. But when you read the paper the "significant risk" only means p < 0.05, and the explained variance was only about 1 % because they had an enourmous sample size. Often yopu ene have to decipher the effect size from the F- or t-values they report. If they had reported the effect size chances are that nobody would be interested.

**Comment by josef-pkt** [comment]
@sturlamolden If you don't like a point hypothesis, use http://statsmodels.sourceforge.net/devel/generated/statsmodels.stats.weightstats.ttost_ind.html

Since this PR enhances a function that calculates a classical hypothesis test, Bayesian alternatives are pretty irrelevant, whether you like p-values or not.

**Comment by sturlamolden** [comment]
I really don't see why those tests are any better than a confidence interval on ```m1 - m2```. Also it computes the p-value which is not valid as "evidence against the null hypothesis" because it violates the likelihood principle. 

A confidence interval has a crazy interpretation, but at least it gives some sort of useful information and does not violate fundamental laws of statistics. Some major medical journals like Journal of Physiology (London) are now even requiring authors to use confidence intervals instead of p-values.

As for a t-test against a non-sharp null hypothesis, in biology or medicine any refree will reject it for publication because they don't find it in their textbook. So you can just forget about getting something like that published. You will be ordred to change it to a normal t-test (sharp null hypothesis) or use a confidence interval.

**Comment by josef-pkt** [comment]
Check the medical literature for "equivalence trials" or superiority trials
https://scholar.google.ca/scholar?q=equivalence+trials&btnG=&hl=en&as_sdt=0%2C5&sciodt=0%2C5&cites=13603780389765822253&scipsc=

that's where I got most of my equivalence testing from

**Comment by sturlamolden** [comment]
From the first hit:

*It is important to emphasise that absolute equivalence can never be demonstrated: it is possible only to assert that the true difference is unlikely to be outside a range which depends on the size of the trial, the results of the trial, and the specified probabilities of error. If we have predefined a range of equivalence as an interval from -δ to +δ we can then simply check whether the confidence interval centred on the observed difference lies entirely between δ and +δ. If it does, equivalence is demonstrated; if it does not, there is still room for doubt.*

I don't disagree with this. But I would not use a "test" and produce a "p-value". I would simply compute the confidence interval and then use my clinical training to discuss whether the observed difference in effect size is likely to be clinically relevant.

**Comment by josef-pkt** [comment]
I'm not getting into the p-value discussion again.
About the importance of confidence intervals (as complement to p-values)
see above https://github.com/scipy/scipy/issues/4399#issuecomment-70411036
I'm all in favor and I'm still waiting for some "effect size" contributions. I want it all, p-values, confidence intervals, effect size and power!

**Comment by mortonjt** [comment]
I won't be able to work on these permutation tests for the next few days.  Once I free up, I'll move on to commit this to statsmodels and incorporate these features.  Sounds good?

**Comment by cilia** [comment]
hey guys, is there any plan for merging these permutation tests in the near future? @mortonjt what is the status of these permutation tests on your branch? Are they stable enough to be used if a merge into scipy is not currently planned? Thanks.

**Comment by mortonjt** [comment]
The unit tests for the permutative t-test / f-test should be well tested -
I've been using these techniques for my own research for a while.

I'm not entirely sure what needs to be finalized before merging the t-test
...

@rgommers, could you provide any insight about what needs to be done before
merging, or if these permutation tests still fit within the scope of scipy?

**Comment by fowles** [comment]
This appears to have stalled.  Any hope of getting it moving again?

**Comment by sturlamolden** [comment]
The performance problems with ```numpy.random.shuffle``` has been fixed upstream. We do not any longer need special Cython shuffling code to make fast permutation tests.

Fast permutation tests was implemented in #4440 which is now superseeded by #4824
